### PR TITLE
[TG-496] Add Goto function unit test utilities

### PR DESCRIPTION
**Pull request by thk123**
Added some unit test utility functions for verifying the contents of a goto function.

### CODE REVIEW

File: src/goto-programs/initialize_goto_model.h [hunk 1]
```diff
@@ -14,11 +14,18 @@ Author: Daniel Kroening, kroening@kroening.com
...
 #include <util/cmdline.h>
+#include <util/language.h>
 
 #include "goto_model.h"
 
 goto_modelt initialize_goto_model(
   const cmdlinet &cmdline,
   message_handlert &message_handler);
 
+bool initialize_goto_model(
```
Comment (thk123) [inline src/goto-programs/initialize_goto_model.h]: Remove this abandoned overload

File: src/java_bytecode/java_bytecode_language.cpp [hunk 2]
```diff
@@ -251,7 +251,7 @@ bool java_bytecode_languaget::typecheck(
     return res;
   }
 
-  instantiate_generics(get_message_handler(), symbol_table);
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_language.cpp]: Remove this properly

File: unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp [hunk 3]
```diff
@@ -9,6 +9,7 @@
 
 #include <testing-utils/catch.hpp>
 #include <testing-utils/load_java_class.h>
+#include <testing-utils/require_goto_statements.h>
```
Comment (thk123) [inline unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp]: Remove unused includes

File: unit/testing-utils/load_java_class.cpp [hunk 4]
```diff
@@ -64,4 +74,4 @@ symbol_tablet load_java_class(
   // as this often indicates that one of these is wrong.
   REQUIRE_FALSE(class_type.get_bool(ID_incomplete_class));
   return new_symbol_table;
-}
+}
```
Comment (thk123) [inline unit/testing-utils/load_java_class.cpp]: Empty last line

File: unit/testing-utils/load_java_class.h [hunk 5]
```diff
@@ -14,9 +14,17 @@
...
 
 symbol_tablet load_java_class(
   const std::string &java_class_name,
   const std::string &class_path);
 
+symbol_tablet load_java_class(
+  const std::string &java_class_name,
+  const std::string &class_path,
+  std::unique_ptr<languaget> java_lang);
+
```
Comment (thk123) [inline unit/testing-utils/load_java_class.h:28]: Nit: remove unnecessary blank lines

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 6]
```diff
@@ -126,46 +126,3 @@ irep_idt generate_java_generic_typet::build_generic_tag(
...
-
-/// Activate the generic instantiation code.
-/// \param message_handler
-/// \param symbol_table The symbol table so far.
-void
-instantiate_generics(
-  message_handlert &message_handler,
-  symbol_tablet &symbol_table)
-{
-  generate_java_generic_typet instantiate_generic_type(message_handler);
```
Comment (smowton) [inline src/java_bytecode/generate_java_generic_type.cpp]: Is all this supposed to be in this PR?

File: src/java_bytecode/generate_java_generic_type.h [hunk 7]
```diff
@@ -31,8 +31,4 @@ class generate_java_generic_typet
   message_handlert &message_handler;
 };
 
-void instantiate_generics(
-  message_handlert &message_handler,
-  symbol_tablet &symbol_table);
```
Comment (tautschnig) [inline src/java_bytecode/generate_java_generic_type.h]: How does a removal of a function fit the PR subject?

File: unit/testing-utils/require_goto_statements.cpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    if(assignment.get_statement() == ID_assign)
+    {
+      const code_assignt &code_assign = to_code_assign(assignment);
+
+      if(code_assign.lhs().id() == ID_member)
+      {
+        const auto &member_expr = to_member_expr(code_assign.lhs());
+        const auto &symbol = member_expr.symbol();
+
+        if(symbol.get_identifier()==structure_name&&
```
Comment (smowton) [inline unit/testing-utils/require_goto_statements.cpp]: space before &&

File: unit/testing-utils/require_goto_statements.cpp [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+/// \param instructions: The value of the function (e.g. got by looking up
+///  the function in the symbol table and getting the value)
+/// \return: All ID_code statements in the tree rooted at \p function_value
+std::vector<codet> require_goto_statements::get_all_statements(
+  const exprt &function_value)
+{
+  std::vector<codet> statements;
+
+  // Add any child statements (e.g. if this is a code_block
+  // TODO(tkiley): It should be possible to have a custom version of
```
Comment (smowton) [inline unit/testing-utils/require_goto_statements.cpp]: This is nasty enough to do with `std` algorithms, I'd just use the for-loop version if I were you.

**Review by smowton — COMMENTED** [review]
Looks ok except I'm not sure why the changes to java_bytecode are here

File: src/java_bytecode/generate_java_generic_type.h (hunk 7 above)
Comment (thk123) [reply src/java_bytecode/generate_java_generic_type.h]: Good point - will attempt to think of a better name for the PR.

File: unit/testing-utils/require_goto_statements.cpp (hunk 9 above)
Comment (thk123) [reply unit/testing-utils/require_goto_statements.cpp]: And just remove the TODO?

Comment (thk123) [reply unit/testing-utils/require_goto_statements.cpp]: You're quite right, for loop looks much simpler

File: unit/testing-utils/require_goto_statements.cpp [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+\*******************************************************************/
+
+#include "require_goto_statements.h"
+#include "catch.hpp"
+
+#include <algorithm>
+#include <util/expr_iterator.h>
+
+/// Expand value of a function to include all child codets
+/// \param instructions: The value of the function (e.g. got by looking up
```
Comment (mgudemann) [inline unit/testing-utils/require_goto_statements.cpp]: parameter has different name

File: src/java_bytecode/generate_java_generic_type.h (hunk 7 above)
Comment (tautschnig) [reply src/java_bytecode/generate_java_generic_type.h]: Could it please just be two different PRs?

File: unit/testing-utils/require_goto_statements.cpp [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        {
+          component_assignments.push_back(code_assign);
+        }
+      }
+    }
+  }
+  return component_assignments;
+}
+
+/// For a given variable name, gets the assignments to it in the functions
```
Comment (mgudemann) [inline unit/testing-utils/require_goto_statements.cpp]: why `functions` not `instructions`?

File: unit/testing-utils/require_goto_statements.cpp [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    }
+  }
+
+  return locations;
+}
+
+/// Find the declaration of the specific variable.
+/// \param variable_name: The name of the variable.
+/// \param entry_point_instructions: The statements to look through
+/// \return The declaration statement corresponding to that variable
```
Comment (mgudemann) [inline unit/testing-utils/require_goto_statements.cpp:113]: please document that it throws an exception when the declaration cannot be found

**Review by mgudemann — CHANGES_REQUESTED** [review]

**Comment by thk123** [comment]
@smowton @tautschnig As requested I have pulled out the removing redundant code into a new branch (PR to follow)

@mgudemann Comments addressed.

**Review by smowton — APPROVED** [review]
The rest is fine with me

**Review by NlightNFotis — APPROVED** [review]
LGTM

**Review by mgudemann — APPROVED** [review]

**Pull request merged**